url3426a97ee4_rtreffer_minidns_tgz-pJ8-de_measite_minidns_minidnsrepl_MiniDnsReplJ8/src/de/measite/minidns/dnssec/DNSSECClient.java

```diff
@@ -0,0 +1,555 @@
+        if (records.isEmpty()) return records;
+        List<Record<? extends Data>> recordList = new ArrayList<>(records.size());
+        for (Record<? extends Data> record : records) {
+            if (record.type != TYPE.RRSIG) {
+                recordList.add(record);
+            }
+        }
+        return recordList;
+    }
+
+    @Override
+    protected boolean isResponseCacheable(Question q, DNSMessage dnsMessage) {
+        return super.isResponseCacheable(q, dnsMessage);
+    }
+
+    private Set<UnverifiedReason> verify(DNSMessage dnsMessage) throws IOException {
+        if (!dnsMessage.answerSection.isEmpty()) {
+            return verifyAnswer(dnsMessage);
+        } else {
+            return verifyNsec(dnsMessage);
+        }
+    }
+
+    private Set<UnverifiedReason> verifyAnswer(DNSMessage dnsMessage) throws IOException {
+        Question q = dnsMessage.questions.get(0);
+        List<Record<? extends Data>> answers = dnsMessage.answerSection;
+        List<Record<? extends Data>> toBeVerified = dnsMessage.copyAnswers();
+        VerifySignaturesResult verifiedSignatures = verifySignatures(q, answers, toBeVerified);
+        Set<UnverifiedReason> result = verifiedSignatures.reasons;
+        if (!result.isEmpty()) {
+            return result;
+        }
+
+        // Keep SEPs separated, we only need one valid SEP.
+        boolean sepSignatureValid = false;
+        Set<UnverifiedReason> sepReasons = new HashSet<>();
+        for (Iterator<Record<? extends Data>> iterator = toBeVerified.iterator(); iterator.hasNext(); ) {
+            Record<DNSKEY> record = iterator.next().ifPossibleAs(DNSKEY.class);
+            if (record == null) {
+                continue;
+            }
+
+            // Verify all DNSKEYs as if it was a SEP. If we find a single SEP we are safe.
+            Set<UnverifiedReason> reasons = verifySecureEntryPoint(q, record);
+            if (reasons.isEmpty()) {
+                sepSignatureValid = true;
+            } else {
+                sepReasons.addAll(reasons);
+            }
+            if (!verifiedSignatures.sepSignaturePresent) {
+                LOGGER.finer("SEP key is not self-signed.");
+            }
+            iterator.remove();
+        }
+
+        if (verifiedSignatures.sepSignaturePresent && !sepSignatureValid) {
+            result.addAll(sepReasons);
+        }
+        if (verifiedSignatures.sepSignatureRequired && !verifiedSignatures.sepSignaturePresent) {
+            result.add(new NoSecureEntryPointReason(q.name.ace));
+        }
+        if (!toBeVerified.isEmpty()) {
+            if (toBeVerified.size() != answers.size()) {
+                throw new DNSSECValidationFailedException(q, "Only some records are signed!");
+            } else {
+                result.add(new NoSignaturesReason(q));
+            }
+        }
+        return result;
+    }
+
+    private Set<UnverifiedReason> verifyNsec(DNSMessage dnsMessage) throws IOException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 13. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 238-247 into a separate method.